java/ql/src/experimental/Security/CWE/CWE-347/Auth0NoVerifierLocalSource.ql

```diff
@@ -0,0 +1,67 @@
+/**
+ * @name Missing JWT signature check
+ * @description Failing to check the Json Web Token (JWT) signature may allow an attacker to forge their own tokens.
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 7.8
+ * @precision medium
+ * @id java/missing-jwt-signature-check-auth0-local-source
+ * @tags security
+ *       external/cwe/cwe-347
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import JwtAuth0 as JwtAuth0
+
+module JwtDecodeConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) {
+    exists(Variable v |
+      source.asExpr() = v.getInitializer() and
+      v.getType().hasName("String")
+    ) and
```

Comment:
This is still a bad source. If your source is just "any string variable" then there is essentially no point in having a source at all. In this case just track flow starting at the decode operation.

Reply:
(on the other hand is there an actual source that we don't call a RemoteFlowSource but ought to be considered dangerous in order to get the CVE hit you're looking for? If so can you say what it is?)

Reply:
I think you should just consider any JWT token potentially interesting, since they're inherently things used in a security-sensitive context. I would recommend making the result of the decode method your source, then just measuring flow from there to either verification or to information access, rather than worry about the ultimate flow source being remote or otherwise.

Reply:
Ah yes I see. I'm not familiar enough with Shiro to know if `doGetAuthenticationInfo` is a reasonable method to assume is generally called with an untrusted parameter. If so, I would be fine with adding as a source, any method implementing `AuthenticatingRealm.doGetAuthenticationInfo` (note Authenticating not Authorizing; Authenticating defines the interface and Authorizing inherits it). Use `Method.overrides+` to find implementations, like

```
exists(Method base, Method override | base.hasQualifiedName("package", "Class", "methodName") and override.overrides+(base))
```

Reply:
Oh or better yet, add it as an experimental models-as-data source, like you did in https://github.com/github/codeql/pull/16708 but defining a `sourceModel` not a `sinkModel`.

Reply:
Sounds good to me-- if you go ahead and add them as sources, we'll be able to get this PR merged